text/0000-render-element-modifiers.md

```diff
@@ -0,0 +1,570 @@
+
+    if (alert) {
+      alert.classList.add('fade-in');
+    }
+  }
+});
+```
+
+After:
+
+```hbs
+{{#if shouldShow}}
+  <div {{did-render this.fadeIn}} class="alert">
+    {{yield}}
+  </div>
+{{/if}}
+```
+```js
+export default Component.extend({
+  fadeIn(element) {
+    element.classList.add('fade-in');
+  }
+});
+```
+
+#### Example: Resizing text area
+
+One key thing to know about `{{did-render}}` is it will not rerun whenever the
+_contents_ or _attributes_ on the element change. For instance, `{{did-render}}`
+will _not_ rerun when `@type` changes here:
+
+```hbs
+<div {{did-render this.setupType}} class={{@type}}></div>
+```
+
+If `{{did-render}}` should rerun whenever a value changes, the value should be
+passed as a parameter to the modifier. For instance, a textarea which wants to
+resize itself to fit text whenever the text is modified could be setup like
+this:
+
+```hbs
+<textarea {{did-render this.resizeArea @text}}>
+  {{@text}}
+</textarea>
+```
+```js
+export default Component.extend({
+  resizeArea(element) {
+    element.css.height = `${element.scrollHeight}px`;
+  }
+});
+```
+
+#### Example: `ember-composability-tools` style rendering
+
+This is the type of rendering done by libraries like `ember-leaflet`, which use
+components to control the _rendering_ of the library, but without any templates
+themselves. The underlying library for this is [here](https://github.com/miguelcobain/ember-composability-tools).
+This is a simplified example of how you could accomplish this with Glimmer
+components and element modifiers.
+
+Node component:
+
+```js
+// components/node.js
+export default Component.extend({
+  init() {
+    super(...arguments);
+    this.children = new Set();
+
+    this.args.parent.registerChild(this);
+  }
+
+  willDestroy() {
+    super(...arguments);
+
+    this.args.parent.unregisterChild(this);
+  }
+
+  registerChild(child) {
+    this.children.add(child);
+  }
+
+  unregisterChild(child) {
+    this.children.delete(child);
+  }
+
+  didInsertNode(element) {
+    // library setup code goes here
+
+    this.children.forEach(c => c.didInsertNode(element));
+  }
+
+  willDestroyNode(element) {
+    // library teardown code goes here
+
+    this.children.forEach(c => c.willDestroyNode(element));
+  }
+}
+```
+```hbs
+<!-- components/node.hbs -->
+{{yield (component "node" parent=this)}}
+```
+
+Root component:
+
+```js
+// components/root.js
+import NodeComponent from './node.js';
+
+export default NodeComponent.extend();
+```
+```hbs
+<!-- components/root.hbs -->
+<div
+  {{did-render (action this.didInsertNode)}}
+  {{will-destroy (action this.willDestroyNode)}}
+>
+  {{yield (component "node" parent=this)}}
+</div>
+```
+
+Usage:
+
+```hbs
+<Root as |node|>
+  <node as |node|>
+    <node />
+  </node>
+</Root>
+```
+
+## Drawbacks
+
+* Element modifiers are a new concept that haven't been fully stabilized as of
+  yet. It may be premature to add default modifiers to the framework.
```

Comment:
FWIW, I think the only way to make them "main stream" is to motivate their use cases which I think this RFC does very well!